File: LeetCode/938-range-sum-of-bst.py
###############################################################################################
# easy
############# 
# Time Complexity: O(n)
# Space Complexity: O(n), the worst is O(n)
###############################################################################################
class Solution:
    def rangeSumBST(self, root: Optional[TreeNode], low: int, high: int) -> int:
        def dfs(node):
            res = 0
            if low <= node.val <= high:
                res += node.val
            if node.left and node.val > low:
                res += dfs(node.left)
            if node.right and node.val < high:
                res += dfs(node.right)
            return res
        return dfs(root)

File: LeetCode/938-range-sum-of-bst.cpp
///////////////////////////////////////////////////////////////////////////////////////////////
// easy
////////////////
// Time Complexity: O(n)
// Space Complexity: O(logn)
///////////////////////////////////////////////////////////////////////////////////////////////
class Solution {
public:
    int rangeSumBST(TreeNode* root, int low, int high) {
        if(!root || low > high) return 0;
    
        int left = rangeSumBST(root->left, low, high);
        int right = rangeSumBST(root->right, low, high);
        
        return ((root->val<=high&&root->val>=low)?root->val:0) + left + right;
    }
};